Scheduler rewrite

andyross opened the pull request:
This is working now for simple test cases, though I'm still playing whack-a-mole with sanitycheck to get it ready for merge.   Should be good for early review though.

(Needs a rebase too -- on top of a week-old tree)

Commit:
kernel: Scheduler rewrite

This replaces the existing scheduler (but not priority handling)
implementation with a somewhat simpler one.  Behavior as to thread
selection does not change.  New features:

+ Unifies SMP and uniprocessing selection code (with the sole
  exception of the "cache" trick not being possible in SMP).

+ The old static multi-queue implementation is gone and has been
  replaced with a build-time choice of either a "dumb" list
  implementation (faster and significantly smaller for apps with only
  a few threads) or a balanced tree queue which scales well to
  arbitrary numbers of threads and priority levels.  This is
  controlled via the CONFIG_SCHED_DUMB kconfig variable.

+ The balanced tree implementation is usable symmetrically for the
  wait_q abstraction, fixing a scalability glitch Zephyr had when many
  threads were waiting on a single object.  This can be selected via
  CONFIG_WAITQ_FAST.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

carlescufi commented:
CC @pizi-nordic

andrewboie commented on kernel/sched.c:
```diff
@@ -35,201 +40,205 @@ static inline int _is_preempt(struct k_thread *thread)
- * places where a thread is taken off the ready queue.
- *
- * Interrupts must be locked when calling this function.
+/* FIXME: this API is glitchy when used in SMP.  If the thread is
```
I think we need to file a bug in GH on this.

andrewboie commented on kernel/sched.c:
```diff
@@ -35,201 +40,205 @@ static inline int _is_preempt(struct k_thread *thread)
-	__ASSERT(abs_prio < K_NUM_PRIORITIES, "prio out-of-range\n");
+	/* The timeout handling is currently synchronized external to
+	 * the scheduler using the legacy global lock.  Should fix
+	 * that.
```
I think we need to file a bug in GH on this.

andrewboie commented on kernel/sched.c:
```diff
@@ -1,29 +1,34 @@
-{
-	return _is_thread_state_set(thread, _THREAD_DUMMY);
-}
+/* Semi-experimental spin locking API using block scope.  Works great
```
I think the final version of this patch should have more confident language than "semi-experimental".
I don't have an opinion on whether this needs to be moved to spinlock.h I'm fine with it here.

pizi-nordic commented on include/kernel.h:
```diff
@@ -376,7 +377,17 @@ struct __thread_entry {
 
 	/* this thread's entry in a ready/wait queue */
-	sys_dnode_t k_q_node;
+	union {
```
IMHO usage of qnode_dump in outside ready queue context (mailbox.c and pipe.c) calls for better name for this field. What about:

```
union
{
sys_dlist_t k_q_node; // Generic queue node used to put the thread on the list.
struct rbnode k_q_node_rb; // Node for ready queue when CONFIG_WAITQ_FAST is used.
} 
```

andyross replied:
The direct queuing in mailbox and pipe is IMHO a pretty bad layering violation and needs to be removed in favor of some other mechanism (maybe extending the new priq abstraction a tiny bit to allow for simple list management).  IPC tools shouldn't be dealing with scheduler internals.  So I agree, but I'd rather fix it on the other side. :)

But FWIW: the "dumb" in that name isn't an informal or internal thing. It's the "dumb list" variant of thread queueing, and as it has scaling impact I thought it should be explicit.  But if it bothers you strongly I can rename it or otherwise hide the "dumb" (behind a more abstracted struct containing the union, maybe) from kernel.h

Commit:
boards/qemu_x86: Enable fast scheduler options

This target is already using the rbtree as part of CONFIG_USERSPACE,
so it incurs no code size overhead (actually it's a little smaller)
when using the scalable scheduler and waitq implementations.

The change also gets us test coverage of those choices on a default CI
target.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Don't _arch_switch() to yourself

The SMP testing missed the case where _Swap() decides to return back
into the _current.  Obviously there is no valid switch handle for the
running thread into which we can restore, and everything blows up.
(What happened is that the new scheduler code opened up a spot where
k_thread_priority_set() does a _reschedule() unconditionally and
doens't check to see whether or not it's needed like the old code).

But that isn't incorrect!  It's entirely possible that _Swap() may
find that no thread is runnable except _current (due, for example, to
another CPU racing the other thread you expected off to sleep or
something).  Don't blow up, check and return a noop.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
lib/rbtree: Fix & document insert comparison order

The scheduler priq implementation was taking advantage of a subtle
behavior of the way the tree presents the order of its arguments (the
node being inserted is always first).  But it turns out the tree got
that wrong in one spot.

As this was subtle voodoo to begin with, it should have been
documented first.  Similarly add a little code to the test case to
guarantee this in the future.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
lib/rbtree: Fix crash condition with empty trees and rb_min/max()

These weren't properly checking the case of an empty tree

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
New version for merge.  Think I have everything working now.

For anyone looking for major deltas from the earlier version:  there aren't many other than code motion and lots of bug fixing.  But the one design change is the "order_key" used in the rbtree priority queue.   See the implementation and comments in _priq_rb_lessthan() and _priq_rb_add().  This is... sort of ugly.  But it's at least tight code and the latency glitch it introduces is no worse than how our existing code would have performed under those circumstances.

andyross commented:
...and of course I pushed the non-rebased tree.  Fixed now.

Commit:
lib/rbtree:  Add RB_FOR_EACH_CONTAINER()

Symmetric API to SYS_DLIST_FOR_EACH_CONTAINER().  Slightly more
typesafe and saves a line of code.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Wait queues aren't dlists anymore

These assertions snuck through in crossed pull requests.  There's a
specific API for _wait_q_t now, you can't hit the list directly
(because it might be a tree).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
samples/subsys/console: Remove unused tx_sem semaphore

This was declared but unused.  And recent toolchains have apparently
started warning on it leading to sanitycheck failures.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel: Bump stack sizes for a few tests on qemu_x86

These tests had very small stacks, and the rbtree scheduler on
qemu_x86 (which does need a little extra stack room, though not much)
is bumping up against the limit.  Increase by ~128 bytes in most
cases.  In the case of the mbox_api test, there are other platforms
(which don't use the tree) which are right against the limit already
and will fail to link with a larger stack, so bump it for qemu_x86
only.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
lib/posix: Port wait_q usage to new API

The pthread mutex changes went in with an adaptation to build with the
new wait queue API, but they did it by using the old dlist hooks
directly through typecasting and union assignment.  That... is sort of
the opposite of the intent to having the new API be abstracted.  The
pthread code worked, but failed once wait queues (on x86) stopped
being dlists.

Simple fix once I saw the problem, anyway.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>
